File: python/03_函数/01_函数基础.py
# ============================================================
# Python 函数 - 基础
# ============================================================

print("=== 1. 函数的定义和调用 ===")
def greet():
    """简单的问候函数"""
    print("Hello, World!")

greet()

print("\n=== 2. 带参数的函数 ===")
def greet_user(name):
    """向指定用户问候"""
    print(f"你好，{name}！")

greet_user("张三")
greet_user("李四")

print("\n=== 3. 带返回值的函数 ===")
def add(a, b):
    """计算两个数的和"""
    return a + b

result = add(3, 5)
print(f"3 + 5 = {result}")

print("\n=== 4. 带默认参数的函数 ===")
def power(base, exponent=2):
    """计算幂，默认是平方"""
    return base ** exponent

print(f"3的平方: {power(3)}")
print(f"3的立方: {power(3, 3)}")

print("\n=== 5. 关键字参数 ===")
def print_info(name, age, city):
    """打印个人信息"""
    print(f"姓名: {name}, 年龄: {age}, 城市: {city}")

print_info("张三", 25, "北京")
print_info(age=30, city="上海", name="李四")  # 顺序可以任意

print("\n=== 6. 可变参数 - *args ===")
def sum_all(*args):
    """计算所有参数的和"""
    total = 0
    for num in args:
        total += num
    return total

print(f"sum_all(1, 2, 3) = {sum_all(1, 2, 3)}")
print(f"sum_all(1, 2, 3, 4, 5) = {sum_all(1, 2, 3, 4, 5)}")

print("\n=== 7. 可变关键字参数 - **kwargs ===")
def print_dict(**kwargs):
    """打印所有关键字参数"""
    for key, value in kwargs.items():
        print(f"  {key}: {value}")

print("个人信息:")
print_dict(name="王五", age=28, job="工程师", city="深圳")

print("\n=== 8. 函数返回多个值 ===")
def calculate(a, b):
    """返回两个数的和、差、积、商"""
    return a + b, a - b, a * b, a / b

add_r, sub_r, mul_r, div_r = calculate(10, 3)
print(f"10 和 3 的计算结果:")
print(f"  和: {add_r}")
print(f"  差: {sub_r}")
print(f"  积: {mul_r}")
print(f"  商: {div_r}")

print("\n=== 9. lambda 表达式（匿名函数） ===")
square = lambda x: x ** 2
print(f"5的平方（lambda）: {square(5)}")

add_lambda = lambda a, b: a + b
print(f"3+4（lambda）: {add_lambda(3, 4)}")

print("\n=== 10. 函数作为参数传递 ===")
nums = [1, 2, 3, 4, 5]
squared = list(map(lambda x: x ** 2, nums))
print(f"原列表: {nums}")
print(f"平方后: {squared}")

File: c/03_函数/01_函数基础.c
/* ============================================================
 * C 语言函数
 * 编译运行：gcc 01_函数基础.c -o 01_函数基础 && ./01_函数基础
 * ============================================================ */

#include <stdio.h>

// 函数声明（函数原型）
void greet();
void greet_user(char name[]);
int add(int a, int b);
int power(int base, int exponent);
int sum_all(int numbers[], int count);
void swap(int *a, int *b);
int factorial(int n);

int main() {
    printf("=== 1. 函数的定义和调用 ===\n");
    greet();

    printf("\n=== 2. 带参数的函数 ===\n");
    greet_user("张三");
    greet_user("李四");

    printf("\n=== 3. 带返回值的函数 ===\n");
    int result = add(3, 5);
    printf("3 + 5 = %d\n", result);

    printf("\n=== 4. 函数声明与定义分离 ===\n");
    printf("函数原型告诉编译器函数的签名\n");
    printf("实际的函数定义可以在文件的其他位置\n");

    printf("\n=== 5. 幂函数 ===\n");
    printf("3的平方: %d\n", power(3, 2));
    printf("3的立方: %d\n", power(3, 3));

    printf("\n=== 6. 数组作为参数 ===\n");
    int nums[] = {1, 2, 3, 4, 5};
    int total = sum_all(nums, 5);
    printf("数组元素和: %d\n", total);

    printf("\n=== 7. 指针参数（交换变量） ===\n");
    int x = 10, y = 20;
    printf("交换前: x = %d, y = %d\n", x, y);
    swap(&x, &y);
    printf("交换后: x = %d, y = %d\n", x, y);

    printf("\n=== 8. 递归函数 ===\n");
    printf("阶乘:\n");
    for (int i = 0; i < 6; i++) {
        printf("%d! = %d\n", i, factorial(i));
    }

    printf("\n=== 9. 函数的作用域 ===\n");
    // 局部变量只能在函数内访问
    int local_var = 100;
    printf("局部变量 local_var = %d\n", local_var);

    printf("\n=== 10. 静态变量 ===\n");
    void counter() {
        static int count = 0; // 静态变量只初始化一次
        count++;
        printf("count = %d\n", count);
    }
    counter();
    counter();
    counter();

    printf("\n=== 11. 全局变量 ===\n");
    // 全局变量在所有函数中都可以访问
    printf("全局变量可以在所有函数中使用，但要谨慎使用\n");

    printf("\n=== 12. 函数指针 ===\n");
    int (*add_ptr)(int, int) = add; // 函数指针
    printf("通过函数指针调用 add: %d\n", add_ptr(3, 5));

    printf("\n=== 13. 头文件和多文件编译说明 ===\n");
    printf("大型项目中，通常:\n");
    printf("  - .h 文件: 存放函数声明\n");
    printf("  - .c 文件: 存放函数定义\n");
    printf("  - 编译: gcc main.c utils.c -o program\n");

    return 0;
}

// 函数定义

void greet() {
    printf("Hello, World!\n");
}

void greet_user(char name[]) {
    printf("你好，%s！\n", name);
}

int add(int a, int b) {
    return a + b;
}

int power(int base, int exponent) {
    int result = 1;
    for (int i = 0; i < exponent; i++) {
        result *= base;
    }
    return result;
}

int sum_all(int numbers[], int count) {
    int total = 0;
    for (int i = 0; i < count; i++) {
        total += numbers[i];
    }
    return total;
}

void swap(int *a, int *b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

int factorial(int n) {
    if (n == 0 || n == 1) {
        return 1;
    }
    return n * factorial(n - 1);
}
